aptos-move/block-executor/src/cold_validation.rs

```diff
@@ -0,0 +1,1015 @@
+        pending_reqs.push(PendingRequirement {
+            requirements,
+            from_idx: calling_txn_idx + 1,
+            to_idx: min_not_scheduled_idx,
+        });
+
+        // Updates to atomic variables while recording pending requirements occur under the
+        // pending_requirements lock to ensure atomicity versus draining to activate.
+        // However, for simplicity and simpler invariants, all updates (including in
+        // finish_validation_requirement) are under the same lock.
+        let _ = self.dedicated_worker_id.compare_exchange(
+            u32::MAX,
+            worker_id,
+            Ordering::Relaxed,
+            Ordering::Relaxed,
+        );
+        let prev_min_idx = self
+            .min_idx_with_unscheduled_requirement
+            .swap(calling_txn_idx + 1, Ordering::Relaxed);
+        if prev_min_idx <= calling_txn_idx {
+            // Record may not be called with with a calling_txn_idx that is higher
+            // or equal to the min_from_idx, as committing calling_txn_idx is impossible
+            // before the pending requirements with lower min index have not been processed
+            // and then any (lower or equal) required validations have not been performed.
+            return Err(code_invariant_error(format!(
+                "Recording validation requirements, min idx = {} <= calling_txn_idx = {}",
+                prev_min_idx, calling_txn_idx
+            )));
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn is_dedicated_worker(&self, worker_id: u32) -> bool {
+        self.dedicated_worker_id.load(Ordering::Relaxed) == worker_id
+    }
+
+    /// If the caller is the dedicated worker, this method:
+    /// (1) Clears the pending requirements: certain statuses (of executing txns) are marked
+    /// for validation after finishing the execution, and the others that may be affected
+    /// (ones that are executed) are transformed into active requirements.
+    /// (2) Returns the highest priority (lowest idx) active requirement as long as it is
+    /// below the idx_threshold. The reason requirements are drained one by one is (a) for
+    /// simplicity, and (b) to allow unblocking txns for commit as soon as possible.
+    /// Note: The caller may prefer to check is_dedicated_worker before calling this method
+    /// to avoid computing the idx_threshold.
+    pub(crate) fn get_validation_requirement<'a>(
+        &self,
+        worker_id: u32,
+        idx_threshold: TxnIndex,
+        statuses: &ExecutionStatuses,
+    ) -> Result<Option<(TxnIndex, Incarnation, ValidationRequirement<'a, R>)>, PanicError> {
+        if !self.is_dedicated_worker(worker_id) {
+            return Ok(None);
+        }
+
+        self.activate_pending_requirements(statuses)?;
+        // Double check as if pending requirements were empty and no new requirements were
+        // activated, the dedicated worker id would be reset.
+        if !self.is_dedicated_worker(worker_id) {
```

Comment:
Why do you have this if twice?

Reply:
Did you follow the comment before? Basically the previous call can reset worker id, in which case we should return

Reply:
clarified the comment a bit better